File: 279_perfect_squares.rb
# @param {Integer} n
# @return {Integer}
def num_squares(n, cache={0 => 0, 1 => 1, 2 => 2, 3 => 3})
  return cache[n] if cache[n]
  
  num = 4
  while num <= n
    cache[num] = num
    
    i = 1
    while i*i <= num
      cache[num] = [cache[num], 1 + cache[num-i*i]].min
      i += 1
    end
    
    num += 1
  end 
  
  cache[n]
end

File: 279_perfect_squares.js
/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function (n) {
  let cache = { 0: 0, 1: 1, 2: 2, 3: 3 };
  cache[n] = numsSquaresHelper(n, cache);
  return cache[n];
};

function numsSquaresHelper(n, cache) {
  if (n <= 0) {
    return 0;
  }

  if (cache[n]) {
    return cache[n];
  }

  let num = 4;

  while (num <= n) {
    cache[num] = num;

    for (let i = 1; i * i <= num; i++) {
      cache[num] = Math.min(cache[num], 1 + cache[num - i * i]);
    }

    num++;
  }

  return cache[n]
}